File: python/hash_table.py
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.next = None


class HashTable:
    def __init__(self, capacity=16):
        self.capacity = capacity
        self.size = 0
        self.table = [None] * capacity

    def index(self, key):
        return hash(key) % self.capacity

    def set(self, key, val):
        index = self.index(key)
        if self.table[index] is None:
            self.table[index] = Node(key, val)
            self.size += 1
        else:
            current = self.table[index]
            while current:
                if current.key == key:
                    return
                current = current.next
            new_node = Node(key, val)
            new_node.next = self.table[index]
            self.table[index] = new_node
            self.size += 1

    def get(self, key):
        index = self.index(key)
        current = self.table[index]
        while current:
            if current.key == key:
                return current.val
            current = current.next
        raise KeyError(key) 

    def remove(self, key):
        index = self.index(key)
        previous = None
        current = self.table[index]

        while current:
            if current.key == key:
                if previous:
                    previous.next = current.next
                else:
                    self.table[index] = current.next
                self.size -= 1
                return       
            previous = current
            current = current.next
        raise KeyError(key)

    def __str__(self):
        elements = []
        for i in range(self.capacity):
            current = self.table[i]
            while current:
                elements.append((current.key, current.val))
                current = current.next
        return str(elements)
        
    def __len__(self):
        return self.size

    def __contains__(self, key):
        try:
            self.get(key)
            return True
        except KeyError:
            return False

File: java/algos-project/src/main/java/uni/algos/hash_table/HashTable.java
package uni.algos.hash_table;

import uni.algos.AbstractMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.AbstractMap.SimpleEntry;

public class HashTable<K, V> extends AbstractMap<K, V> {
  private static class Node<K, V> extends SimpleEntry<K, V> {
    private Node<K, V> next;

    Node(K key, V value, Node<K, V> next) {
      super(key, value);
      this.next = next;
    }
  }

  private static final int INITIAL_CAPACITY = 16; // Starting size of the table
  private static final float LOAD_FACTOR = 0.75f; // Load factor for resizing

  private Node<K, V>[] table; // Array of nodes (buckets)
  private int size = 0; // Number of elements in the hash table

  // Constructor to initialize the hash table
  public HashTable() {
    table = new Node[INITIAL_CAPACITY];
  }

  // Put method to add or update a key-value pair
  @Override
  public V put(K key, V value) {
    int index = getIndex(key);

    // If no node exists at the index, create one
    if (table[index] == null) {
      table[index] = new Node<>(key, value, null);
    } else {
      Node<K, V> current = table[index];
      while (current.next != null) {
        if (current.getKey().equals(key)) {
          current.setValue(value); // Update value if key already exists
          return value;
        }
        current = current.next;
      }
      current.next = new Node<>(key, value, null); // Add to the end
    }

    size++;

    // Resize if load factor exceeded
    if (size > LOAD_FACTOR * table.length) {
      resize();
    }
    return value;
  }

  // Get method to retrieve the value associated with a key
  @Override
  public V get(Object key) {
    int index = getIndex(key);
    Node<K, V> current = table[index];

    while (current != null) {
      if (current.getKey().equals(key)) {
        return current.getValue();
      }
      current = current.next;
    }
    return null; // Key not found
  }

  // Remove method to delete a key-value pair
  @Override
  public V remove(Object key) {
    int index = getIndex(key);
    Node<K, V> current = table[index];
    Node<K, V> previous = null;

    while (current != null) {
      if (current.getKey().equals(key)) {
        if (previous == null) {
          table[index] = current.next; // Remove the head node
        } else {
          previous.next = current.next; // Remove the current node
        }
        size--;
        return previous.getValue();
      }
      previous = current;
      current = current.next;
    }
    return null;
  }

  protected int hash(Object object) {
    return object.hashCode();
  }

  protected int getIndex(Object key) {
    int hash = hash(key);
    return (hash & 0x7FFFFFFF) % table.length; // Ensure non-negative index
  }

  private void resize() {
    Node<K, V>[] oldTable = table;
    table = new Node[oldTable.length * 2];
    size = 0;

    for (Node<K, V> headNode : oldTable) {
      Node<K, V> current = headNode;
      while (current != null) {
        put(current.getKey(), current.getValue());
        current = current.next;
      }
    }
  }

  @Override
  public int size() {
    return size;
  }
}
